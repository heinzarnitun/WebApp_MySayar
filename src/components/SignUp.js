import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { register } from '../services/authService';
import axios from 'axios';
import API_URL from '../api';

const SignUp = () => {
  const [userData, setUserData] = useState({
    name: '',
    email: '',
    password: '',
    password_confirmation: '',
  });
  const [errors, setErrors] = useState({}); // Store validation errors
  const [success, setSuccess] = useState('');
  const navigate = useNavigate();

  const handleChange = (e) => {
    setUserData({ ...userData, [e.target.name]: e.target.value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    setErrors({}); // Reset errors before submission

    try {
      setSuccess('Registration successful! Redirecting to login...');
      
      setTimeout(() => {
        navigate('/sign-in');
      }, 1000);
      // Fetch CSRF token before making a request
      await axios.get(`${API_URL}/sanctum/csrf-cookie`);

      // Call the register function from authService
      await register(userData);

       // Redirect after successful registration
    } catch (error) {
      if (error.response && error.response.data.errors) {
        // Set Laravel's validation error messages
        setErrors(error.response.data.errors);
      } else {
        setErrors({ general: 'Registration failed: Server error' });
      }
    }
  };

  return (
    <div className="container my-5">
      <div className="row justify-content-center">
        <div className="col-md-8 col-lg-6">
          <div className="card shadow">
            <div className="card-body p-4 p-md-5">
              
              {/* Header */}
              <div className="text-center mb-4">
                <h1 className="fw-bold mb-3">Create Account</h1>
                {success && (
                  <div className="alert alert-success" role="alert">
                    {success}
                  </div>
                )}
              </div>

              {/* Form */}
              <form onSubmit={handleSubmit}>
                {/* Name Field */}
                <div className="mb-3">
                  <label htmlFor="name" className="form-label">Full Name</label>
                  <input
                    type="text"
                    className={`form-control ${errors.name ? 'is-invalid' : ''}`}
                    id="name"
                    name="name"
                    placeholder="John Doe"
                    value={userData.name}
                    onChange={handleChange}
                    required
                  />
                  {errors.name && (
                    <div className="invalid-feedback">
                      {errors.name[0]}
                    </div>
                  )}
                </div>

                {/* Email Field */}
                <div className="mb-3">
                  <label htmlFor="email" className="form-label">Email</label>
                  <input
                    type="email"
                    className={`form-control ${errors.email ? 'is-invalid' : ''}`}
                    id="email"
                    name="email"
                    placeholder="your@email.com"
                    value={userData.email}
                    onChange={handleChange}
                    required
                  />
                  {errors.email && (
                    <div className="invalid-feedback">
                      {errors.email[0]}
                    </div>
                  )}
                </div>

                {/* Password Field */}
                <div className="mb-3">
                  <label htmlFor="password" className="form-label">Password</label>
                  <input
                    type="password"
                    className={`form-control ${errors.password ? 'is-invalid' : ''}`}
                    id="password"
                    name="password"
                    placeholder="••••••••"
                    value={userData.password}
                    onChange={handleChange}
                    required
                  />
                  {errors.password && (
                    <div className="invalid-feedback">
                      {errors.password[0]}
                    </div>
                  )}
                </div>

                {/* Confirm Password Field */}
                <div className="mb-4">
                  <label htmlFor="password_confirmation" className="form-label">Confirm Password</label>
                  <input
                    type="password"
                    className={`form-control ${errors.password_confirmation ? 'is-invalid' : ''}`}
                    id="password_confirmation"
                    name="password_confirmation"
                    placeholder="••••••••"
                    value={userData.password_confirmation}
                    onChange={handleChange}
                    required
                  />
                  {errors.password_confirmation && (
                    <div className="invalid-feedback">
                      {errors.password_confirmation[0]}
                    </div>
                  )}
                </div>

                {/* General Errors */}
                {errors.general && (
                  <div className="alert alert-danger mb-4" role="alert">
                    {errors.general}
                  </div>
                )}

                {/* Submit Button */}
                <button 
                  type="submit" 
                  className="btn btn-primary w-100 py-2 fw-bold"
                >
                  Sign Up
                </button>
              </form>

              {/* Footer Links */}
              <div className="mt-4 text-center">
                <p className="text-muted mb-0">
                  Already have an account? <a href="/sign-in" className="text-decoration-none">Sign in</a>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default SignUp;
