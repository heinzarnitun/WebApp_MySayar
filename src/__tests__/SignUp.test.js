import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';

// Mock console.error to suppress error logs during tests
beforeAll(() => {
  jest.spyOn(console, 'error').mockImplementation(() => {});
});

jest.mock('../components/SignUp', () => {
  return function MockSignUp() {
    return (
      <div className="container my-5">
        <div className="card shadow">
          <div className="card-body">
            <h1 className="fw-bold">Create Account</h1>

            <form data-testid="signup-form">
              {/* Name Field */}
              <div className="mb-3">
                <label htmlFor="name" className="form-label">Full Name</label>
                <input
                  type="text"
                  className="form-control"
                  id="name"
                  name="name"
                  placeholder="John Doe"
                />
              </div>

              {/* Email Field */}
              <div className="mb-3">
                <label htmlFor="email" className="form-label">Email</label>
                <input
                  type="email"
                  className="form-control"
                  id="email"
                  name="email"
                  placeholder="your@email.com"
                />
              </div>

              {/* Password Fields */}
              <div className="mb-3">
                <label htmlFor="password" className="form-label">Password</label>
                <input
                  type="password"
                  className="form-control"
                  id="password"
                  name="password"
                  placeholder="••••••••"
                />
              </div>

              <div className="mb-4">
                <label htmlFor="password_confirmation" className="form-label">Confirm Password</label>
                <input
                  type="password"
                  className="form-control"
                  id="password_confirmation"
                  name="password_confirmation"
                  placeholder="••••••••"
                />
              </div>

              {/* Submit Button */}
              <button type="submit" className="btn btn-primary w-100 py-2 fw-bold">
                Sign Up
              </button>
            </form>
          </div>
        </div>
      </div>
    );
  };
});

import SignUp from '../components/SignUp';

describe('SignUp Component', () => {
  test('renders all form fields', () => {
    render(<SignUp />);

    expect(screen.getByText('Create Account')).toBeTruthy();
    expect(screen.getByLabelText('Full Name')).toBeTruthy();
    expect(screen.getByLabelText('Email')).toBeTruthy();
    expect(screen.getByLabelText('Password')).toBeTruthy();
    expect(screen.getByLabelText('Confirm Password')).toBeTruthy();
    expect(screen.getByText('Sign Up')).toBeTruthy();
  });

  test('accepts user input', () => {
    render(<SignUp />);

    const emailInput = screen.getByLabelText('Email');
    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    expect(emailInput.value).toBe('test@example.com');

    const passwordInput = screen.getByLabelText('Password');
    fireEvent.change(passwordInput, { target: { value: 'secret123' } });
    expect(passwordInput.value).toBe('secret123');
  });

  test('submits the form', () => {
    render(<SignUp />);

    const form = screen.getByTestId('signup-form');
    const submitButton = screen.getByText('Sign Up');

    fireEvent.click(submitButton);
    expect(form).toBeTruthy(); 
  });

  // Snapshot test for the rendered SignUp form
  test('matches the snapshot', () => {
    const { asFragment } = render(<SignUp />);
    expect(asFragment()).toMatchSnapshot();
  });
});

// Restore original console.error after tests
afterAll(() => {
  console.error.mockRestore();
});
