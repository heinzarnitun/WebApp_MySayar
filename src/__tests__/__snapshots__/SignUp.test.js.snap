// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container my-5"
  >
    <div
      class="card shadow"
    >
      <div
        class="card-body"
      >
        <h1
          class="fw-bold"
        >
          Create Account
        </h1>
        <form
          data-testid="signup-form"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="name"
            >
              Full Name
            </label>
            <input
              class="form-control"
              id="name"
              name="name"
              placeholder="John Doe"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="email"
            >
              Email
            </label>
            <input
              class="form-control"
              id="email"
              name="email"
              placeholder="your@email.com"
              type="email"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              placeholder="••••••••"
              type="password"
            />
          </div>
          <div
            class="mb-4"
          >
            <label
              class="form-label"
              for="password_confirmation"
            >
              Confirm Password
            </label>
            <input
              class="form-control"
              id="password_confirmation"
              name="password_confirmation"
              placeholder="••••••••"
              type="password"
            />
          </div>
          <button
            class="btn btn-primary w-100 py-2 fw-bold"
            type="submit"
          >
            Sign Up
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
